---
title: "Zeng data"
output: html_notebook
---

This dataset contains very different types of data: 10X cells/nuclei and SmartSeq cells/nuclei.

# 10x data

## Standard pipeline

There are less files compared to Macosko. We only have the molecule_info.h5, all other files are missing. Let us try using cellrangerRkit as we did with the Macosko data:

```{r message=FALSE}
library(cellrangerRkit)
data_dir <- "~/work/data/BICCN_18_05_21/Zeng/10X_cells_MOp/"
test_dir <- "~/work/data/BICCN_18_05_21/Zeng/10X_cells_MOp/L8TX_171026_01_A04/"
```

```{r}
# cellrangerRkit::load_cellranger_matrix(test_dir) ## fails because the MEX files are missing
```

The missing files seem to be important for cellranger to work properly. How can we generate them?

```{r}
h5_path <- "~/work/data/BICCN_18_05_21/Zeng/10X_cells_MOp/L8TX_171026_01_A04/outs/molecule_info.h5"
# load_molecule_info_from_path(test, h5_path) ## memory issues
```
The molecule info cannot be loaded in memory properly. Aslo the 10x genomics side seems to say that the function used to load molecule_info files is deprecated?

## Opening files containing UMIs with rhdf5

Let us try to load the file containing UMIs. Maybe it contains the data from all samples.

```{r}
data_dir <- "~/work/data/BICCN_18_05_21/Zeng/10X_cells_MOp/"
metadata <- read.table(file.path(data_dir, "sample_metadata.csv"), header=TRUE, sep = ",")
```

```{r}
dim(metadata)
head(metadata)
```

So there seems to be 187908 cells if I'm not mistaken. Can we manage to load the UMI data?

```{r}
library(rhdf5)
```

```{r}
h5_path <- "~/work/data/BICCN_18_05_21/Zeng/10X_cells_MOp/umi_counts.h5"
h5ls(h5_path)
```

```{r}
h5_data <- h5read(h5_path, "/mm10_mrna")
```

```{r}
str(h5_data)
```

So it seems that data indeed contains the counts, but it is hard to know how it is organized. Indices probably contains the mapping to genes:

```{r}
max(h5_data$indices)
```

but where is the mapping to samples? Is everythin aggregated here? Probably inptr is useful here. There is a chance that it indicates where a cell starts and ends.

```{r}
head(h5_data$indptr)
tail(h5_data$indptr)
```

I could not be completely sure, but that seems consistent. Because the list starts with zero, I would assume that the range of the ith cell is [inptr[i] + 1, indptr[i+1]].

## Loading molecule_info files with rhdf5

Before we try to make use the UMI data, I would like to know if it can be reextracted in some way from the original molecule_info files. Let's go back to our original example:

```{r}
rm(list = ls())
h5_path <- "~/work/data/BICCN_18_05_21/Zeng/10X_cells_MOp/L8TX_171026_01_A04/outs/molecule_info.h5"
h5ls(h5_path)
```

Hard to say... Let's load the data.

```{r}
# sample_data <- h5read(h5_path, "/") ## fails: not enough memory
```

Ok, so let's actually figure out what is going on based on the output above. There seems to be a lot of low level information... Let us see if we can load each name individually.

```{r}
sample_data <- h5read(h5_path, "/barcode")
```

```{r}
# length(unique(sample_data))
## result is 373228, I don't run it to avoid memory issues
rm(sample_data)
```

So it seems that there is a large number of cells here. That probably means that the UMI file only contains filtered cells. It should be possible to build the gene count matrix from this data, but there is no point doing it, better to just request it.

## Confirming UMI data based on preliminary analyses

According to the UMI files, 133,133 cells were left. According to the summary of analyses, 128,094 cells were considered, of which 118,172 remained after eliminating potential doublets and such.


# SmartSeq data

The data structure looks similar to the DroNc-Seq data. Let's try to use it the same way:

```{r}
rm(list = ls())
data_dir <- "~/work/data/BICCN_18_05_21/Zeng/SMARTer_cells_MOp/exon_counts.csv.gz"
result <- read.table(gzfile(data_dir), header = TRUE, sep = ",")
rownames(result) <- result$sample_id
result <- t(subset(result, select = -sample_id))
```

There are small differences in format, so we will need to use different routines.

```{r}
dim(result)
```

How is it possible that there are so many genes...

```{r}
library(tidyverse)
gene_statistics <- data.frame(number_transcripts = apply(result, 1, sum))
ggplot(gene_statistics) + geom_histogram(aes(number_transcripts)) + scale_x_continuous(trans="log1p")
sum(gene_statistics$number_transcripts == 0)
```

```{r}
sum(gene_statistics$number_transcripts == 0)
```

There are a lot of zero genes, but I am not sure why there are so many still, probably a lot of noncoding RNAs.

```{r}
rownames(result)[25000:25010]
sum(endsWith(rownames(result), "Rik"))
```


Let's see what happens with introns

```{r}
rm(list = ls())
data_dir <- "~/work/data/BICCN_18_05_21/Zeng/SMARTer_cells_MOp/intron_counts.csv.gz"
result <- read.table(gzfile(data_dir), header = TRUE, sep = ",")
rownames(result) <- result$sample_id
result <- t(subset(result, select = -sample_id))
```


```{r}
dim(result)
```

So there is the same number of genes in total.

```{r}
gene_statistics <- data.frame(number_transcripts = apply(result, 1, sum))
ggplot(gene_statistics) + geom_histogram(aes(number_transcripts)) + scale_x_continuous(trans="log1p")
sum(gene_statistics$number_transcripts == 0)
```

